File: Script/src/utility/finalizer.ts
namespace Twinning.Script {

	export class Finalizer {

		// #region constructor

		private m_action: () => void;

		private m_disposed: boolean;

		// ----------------

		public constructor(
			action: () => void,
		) {
			this.m_action = action;
			this.m_disposed = false;
			return;
		}

		// #endregion

		// #region dispose

		public dispose(
		): void {
			assert_test(!this.m_disposed);
			this.m_disposed = true;
			this.m_action();
			return;
		}

		// ----------------

		[Symbol.dispose](
		): void {
			if (!this.m_disposed) {
				this.dispose();
			}
			return;
		}

		// #endregion

	}

}

File: Assistant/lib/utility/finalizer.dart
import '/common.dart';

// ----------------

class Finalizer {

  // #region constructor

  Future<Void> Function() _action;

  // ----------------

  Finalizer(
    Future<Void> Function() action,
  ) :
    this._action = action;

  // #endregion

  // #region dispose

  Future<Void> dispose(
  ) async {
    await this._action();
    return;
  }

  // #endregion

}
